scalding-core/src/main/scala/com/twitter/scalding/Job.scala

```diff
@@ -46,6 +46,86 @@ object Job {
       .newInstance(args)
       .asInstanceOf[Job]
   }
+
+  /**
+   * Make a job reflectively from the given class
+   * and the Args contained in the Config.
+   */
+  def makeJob[J <: Job](cls: Class[J]): Execution[J] =
+    Execution.getConfigMode.flatMap { case (conf, mode) =>
+      // Now we need to allocate the job
+      Execution.from {
+        val argsWithMode = Mode.putMode(mode, conf.getArgs)
+        cls.getConstructor(classOf[Args])
+          .newInstance(argsWithMode)
+      }
+    }
+
+  /**
+   * Create a job reflectively from a class, which handles threading
+   * through the Args and Mode correctly in the way Job subclasses expect
+   */
+  def toExecutionFromClass[J <: Job](cls: Class[J], onEmpty: Execution[Unit]): Execution[Unit] =
+    makeJob(cls).flatMap(toExecution(_, onEmpty))
+
+  /**
+   * Convert Jobs that only use the TypedPipe API to an Execution
+   *
+   * This can fail for some exotic jobs, but for standard subclasses
+   * of Job (that don't override existing methods in Job except config)
+   * it should work
+   *
+   * onEmpty is the execution to run if you have an empty job. Common
+   * values might be Execution.unit or
+   * Execution.failed(new Exeception("unexpected empty execution"))
+   */
+  def toExecution(job: Job, onEmpty: Execution[Unit]): Execution[Unit] =
+    job match {
+      case (exJob: ExecutionJob[_]) => exJob.execution.unit
+      case _ =>
+        val fd = job.flowDef
+        val rfd = new RichFlowDef(fd)
+
+        val ex = if (rfd.isEmpty) {
+          // TypedPipe jobs should not have modified
+          // the FlowDef yet, only the FlowState should
+          // be updated
+          FlowStateMap.get(fd) match { // Note, we want this to be a pure function so we don't mutate the FlowStateMap
+            case Some(FlowState(srcs, confs, writes)) if srcs.isEmpty && confs.isEmpty =>
+              writes match {
+                case Nil => onEmpty
+                case nonEmpty =>
+                  def write[A](f: FlowStateMap.TypedWrite[A]): Execution[Unit] =
+                    f.pipe.writeExecution(f.sink)
+
+                  Execution.sequence(nonEmpty.map(write(_))).unit
```

Comment:
This is where I get a little bit lost - the rest of this looks great. I'm assuming this doesn't handle a case that would need to be a `writeThrough` in `Execution`? If so, I'm missing how, and if not, it's not clear to me what the failure would look like; maybe there could be a test for that?

Reply:
That is already handled by Executions:
```scala
pipe1.writeExecution(sink1)
  .zip(pipe1.map(fn).writeExecution(sink2))
```
will work. Internally the optimizer sees the fork and that is exposed to cascading (they share everything above the fork).

Really, this is just a trick to get things written into Execution form. You question would still apply to Execution if it were an issue.